tabpy-server/tabpy_server/handlers/base_handler.py

```diff
@@ -47,23 +53,230 @@ def _add_CORS_header(self):
+            'for API "{api_version}"')
+        return False, ''
+
+    def _get_basic_auth_credentials(self) -> bool:
+        '''
+        Find credentials for basic access authentication method. Credentials if
+        found stored in self.username and self.password.
+
+        Returns
+        -------
+        bool
+            True if valid credentials were found.
+            False otherwise.
+        '''
+        logger.debug('Checking request headers for authentication data')
+        if 'Authorization' not in self.request.headers:
+            logger.info('Authorization header not found')
+            return False
+
+        auth_header = self.request.headers['Authorization']
+        auth_header_list = auth_header.split(' ')
+        if len(auth_header_list) != 2 or\
+                auth_header_list[0] != 'Basic':
+            logger.error(f'Unknown authentication method "{auth_header}"')
+            return False
+
+        try:
+            cred = base64.b64decode(auth_header_list[1]).decode('utf-8')
+        except (binascii.Error, UnicodeDecodeError) as ex:
+            logger.critical(f'Cannot decode credentials: {str(ex)}')
+            return False
+
+        login_pwd = cred.split(':')
+        if len(login_pwd) != 2:
+            logger.error('Invalid string in encoded credentials')
+            return False
+
+        self.username = login_pwd[0]
+        self.password = login_pwd[1]
+        return True
+
+    def _get_credentials(self, method) -> bool:
+        '''
+        Find credentials for specified authentication method. Credentials if
+        found stored in self.username and self.password.
+
+        Parameters
+        ----------
+        method: str
+            Authentication method name.
+
+        Returns
+        -------
+        bool
+            True if valid credentials were found.
+            False otherwise.
+        '''
+        if method == 'basic-auth':
+            return self._get_basic_auth_credentials()
+        # Add new methods here...
+
+        # No known methods were found
+        logger.critical(
+            f'Unknown authentication method(s) "{method}" are configured '
+            'for API "{api_version}"')
+        return False
+
+    def _validate_basic_auth_credentials(self) -> bool:
+        '''
+        Validates username:pwd if they are the same as
+        stored credentials.
+
+        Returns
+        -------
+        bool
+            True if credentials has key login and
+            credentials[login] equal SHA3(pwd), False
+            otherwise.
+        '''
+        login = self.username.lower()
+        logger.debug(f'Validating credentials for user name "{login}"')
+        if login not in self.credentials:
+            logger.error(f'User name "{self.username}" not found')
+            return False
+
+        hashed_pwd = hash_password(self.username, self.password)
+        if self.credentials[login].lower() != hashed_pwd.lower():
```

Comment:
isn't login already lowercase at this point?

Reply:
and is the hashed password case insensitive?

Reply:
we deal with base64 of hashed password and need the comparison to be case-insensitive as a-f are the same as A-F.